### PR TITLE
Prefer the use of Terraform Structure helper for normalizing JSON

### PR DESCRIPTION
**Pull request by Ninir**
#### Description
This work reduces duplicated code in the provider by using helpers made in the Terraform core repository. This is focusing on the JSON string normalization.

#### Tests

```
$ make testacc TEST=./aws TESTARGS='-run=TestAccDataSourceAwsVpcEndpoint_'

==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run=TestAccDataSourceAwsVpcEndpoint_ -timeout 120m
?   	github.com/terraform-providers/terraform-provider-aws	[no test files]
=== RUN   TestAccDataSourceAwsVpcEndpoint_basic
--- PASS: TestAccDataSourceAwsVpcEndpoint_basic (57.31s)
=== RUN   TestAccDataSourceAwsVpcEndpoint_byId
--- PASS: TestAccDataSourceAwsVpcEndpoint_byId (58.66s)
=== RUN   TestAccDataSourceAwsVpcEndpoint_withRouteTable
--- PASS: TestAccDataSourceAwsVpcEndpoint_withRouteTable (63.06s)

$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSBatchJobDefinition_'
=== RUN   TestAccAWSBatchJobDefinition_basic
--- PASS: TestAccAWSBatchJobDefinition_basic (12.84s)
=== RUN   TestAccAWSBatchJobDefinition_updateForcesNewResource
--- PASS: TestAccAWSBatchJobDefinition_updateForcesNewResource (20.99s)

$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSCloudWatchDashboard_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run=TestAccAWSCloudWatchDashboard_ -timeout 120m
=== RUN   TestAccAWSCloudWatchDashboard_importBasic
--- PASS: TestAccAWSCloudWatchDashboard_importBasic (17.69s)
=== RUN   TestAccAWSCloudWatchDashboard_basic
--- PASS: TestAccAWSCloudWatchDashboard_basic (16.34s)
=== RUN   TestAccAWSCloudWatchDashboard_update
--- PASS: TestAccAWSCloudWatchDashboard_update (27.38s)

$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSCloudWatchEventRule_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run=TestAccAWSCloudWatchEventRule_ -timeout 120m
?   	github.com/terraform-providers/terraform-provider-aws	[no test files]
=== RUN   TestAccAWSCloudWatchEventRule_importBasic
--- PASS: TestAccAWSCloudWatchEventRule_importBasic (14.44s)
=== RUN   TestAccAWSCloudWatchEventRule_basic
--- PASS: TestAccAWSCloudWatchEventRule_basic (21.37s)
=== RUN   TestAccAWSCloudWatchEventRule_full
--- PASS: TestAccAWSCloudWatchEventRule_full (11.35s)
=== RUN   TestAccAWSCloudWatchEventRule_enable
--- PASS: TestAccAWSCloudWatchEventRule_enable (30.27s)

TF_ACC=1 go test ./... -v -run=TestAccAWSEcsTaskDefinition_ -timeout 120m
?   	github.com/terraform-providers/terraform-provider-aws	[no test files]

=== RUN   TestAccAWSEcsTaskDefinition_basic
--- PASS: TestAccAWSEcsTaskDefinition_basic (20.54s)
=== RUN   TestAccAWSEcsTaskDefinition_withScratchVolume
--- PASS: TestAccAWSEcsTaskDefinition_withScratchVolume (13.01s)
=== RUN   TestAccAWSEcsTaskDefinition_withEcsService
--- PASS: TestAccAWSEcsTaskDefinition_withEcsService (59.56s)
=== RUN   TestAccAWSEcsTaskDefinition_withTaskRoleArn
--- PASS: TestAccAWSEcsTaskDefinition_withTaskRoleArn (15.74s)
=== RUN   TestAccAWSEcsTaskDefinition_withNetworkMode
--- PASS: TestAccAWSEcsTaskDefinition_withNetworkMode (16.07s)
=== RUN   TestAccAWSEcsTaskDefinition_constraint
--- PASS: TestAccAWSEcsTaskDefinition_constraint (11.45s)
=== RUN   TestAccAWSEcsTaskDefinition_changeVolumesForcesNewResource
--- PASS: TestAccAWSEcsTaskDefinition_changeVolumesForcesNewResource (21.21s)
=== RUN   TestAccAWSEcsTaskDefinition_arrays
--- PASS: TestAccAWSEcsTaskDefinition_arrays (12.35s)
=== RUN   TestAccAWSEcsTaskDefinition_Fargate
--- PASS: TestAccAWSEcsTaskDefinition_Fargate (12.68s)
=== RUN   TestAccAWSEcsTaskDefinition_ExecutionRole
--- PASS: TestAccAWSEcsTaskDefinition_ExecutionRole (17.66s)

$ TF_ACC=1 go test ./aws -v -run=TestAccAWSVpcEndpoint_ -timeout 120m
=== RUN   TestAccAWSVpcEndpoint_importBasic
--- PASS: TestAccAWSVpcEndpoint_importBasic (60.66s)
=== RUN   TestAccAWSVpcEndpoint_basic
--- PASS: TestAccAWSVpcEndpoint_basic (55.54s)
=== RUN   TestAccAWSVpcEndpoint_withRouteTableAndPolicy
--- PASS: TestAccAWSVpcEndpoint_withRouteTableAndPolicy (109.74s)
=== RUN   TestAccAWSVpcEndpoint_WithoutRouteTableOrPolicyConfig
--- PASS: TestAccAWSVpcEndpoint_WithoutRouteTableOrPolicyConfig (52.26s)
=== RUN   TestAccAWSVpcEndpoint_removed
--- PASS: TestAccAWSVpcEndpoint_removed (48.49s)

=== RUN   TestAccAWSBeanstalkEnv_basic
--- PASS: TestAccAWSBeanstalkEnv_basic (491.00s)
=== RUN   TestAccAWSBeanstalkEnv_tier
--- PASS: TestAccAWSBeanstalkEnv_tier (648.30s)
=== RUN   TestAccAWSBeanstalkEnv_outputs
--- PASS: TestAccAWSBeanstalkEnv_outputs (476.80s)
=== RUN   TestAccAWSBeanstalkEnv_cname_prefix
--- PASS: TestAccAWSBeanstalkEnv_cname_prefix (536.08s)
=== RUN   TestAccAWSBeanstalkEnv_config
--- PASS: TestAccAWSBeanstalkEnv_config (664.85s)
=== RUN   TestAccAWSBeanstalkEnv_resource
--- PASS: TestAccAWSBeanstalkEnv_resource (474.46s)
=== RUN   TestAccAWSBeanstalkEnv_vpc
--- PASS: TestAccAWSBeanstalkEnv_vpc (528.04s)
=== RUN   TestAccAWSBeanstalkEnv_basic_settings_update
--- PASS: TestAccAWSBeanstalkEnv_basic_settings_update (870.96s)
=== RUN   TestAccAWSBeanstalkEnv_version_label
--- PASS: TestAccAWSBeanstalkEnv_version_label (1024.64s)
=== RUN   TestAccAWSBeanstalkEnv_settingWithJsonValue
--- PASS: TestAccAWSBeanstalkEnv_settingWithJsonValue (558.26s)

$ TF_ACC=1 go test ./aws -v -run=TestAccAWSSQSQueue_ -timeout 120m   
=== RUN   TestAccAWSSQSQueue_importBasic
--- PASS: TestAccAWSSQSQueue_importBasic (28.90s)
=== RUN   TestAccAWSSQSQueue_importFifo
--- PASS: TestAccAWSSQSQueue_importFifo (27.69s)
=== RUN   TestAccAWSSQSQueue_importEncryption
--- PASS: TestAccAWSSQSQueue_importEncryption (27.85s)
=== RUN   TestAccAWSSQSQueue_basic
--- PASS: TestAccAWSSQSQueue_basic (61.90s)
=== RUN   TestAccAWSSQSQueue_tags
--- PASS: TestAccAWSSQSQueue_tags (63.64s)
=== RUN   TestAccAWSSQSQueue_namePrefix
--- PASS: TestAccAWSSQSQueue_namePrefix (25.82s)
=== RUN   TestAccAWSSQSQueue_policy
--- PASS: TestAccAWSSQSQueue_policy (34.59s)
=== RUN   TestAccAWSSQSQueue_redrivePolicy
--- PASS: TestAccAWSSQSQueue_redrivePolicy (33.52s)
=== RUN   TestAccAWSSQSQueue_Policybasic
--- PASS: TestAccAWSSQSQueue_Policybasic (33.41s)
=== RUN   TestAccAWSSQSQueue_FIFO
--- PASS: TestAccAWSSQSQueue_FIFO (23.54s)
=== RUN   TestAccAWSSQSQueue_FIFOExpectNameError
--- PASS: TestAccAWSSQSQueue_FIFOExpectNameError (5.64s)
=== RUN   TestAccAWSSQSQueue_FIFOWithContentBasedDeduplication
--- PASS: TestAccAWSSQSQueue_FIFOWithContentBasedDeduplication (24.36s)
=== RUN   TestAccAWSSQSQueue_ExpectContentBasedDeduplicationError
--- PASS: TestAccAWSSQSQueue_ExpectContentBasedDeduplicationError (6.50s)
=== RUN   TestAccAWSSQSQueue_Encryption
--- PASS: TestAccAWSSQSQueue_Encryption (22.93s)

$ TF_ACC=1 go test ./aws -v -run=TestAccAWSSNSTopic_ -timeout 120m
=== RUN   TestAccAWSSNSTopic_importBasic
--- PASS: TestAccAWSSNSTopic_importBasic (23.92s)
=== RUN   TestAccAWSSNSTopic_basic
--- PASS: TestAccAWSSNSTopic_basic (21.75s)
=== RUN   TestAccAWSSNSTopic_policy
--- PASS: TestAccAWSSNSTopic_policy (22.74s)
=== RUN   TestAccAWSSNSTopic_withIAMRole
--- PASS: TestAccAWSSNSTopic_withIAMRole (37.01s)
=== RUN   TestAccAWSSNSTopic_withFakeIAMRole
--- PASS: TestAccAWSSNSTopic_withFakeIAMRole (73.26s)
=== RUN   TestAccAWSSNSTopic_withDeliveryPolicy
--- PASS: TestAccAWSSNSTopic_withDeliveryPolicy (22.33s)

=== RUN   TestAccAWSCloudFormation_importBasic
--- PASS: TestAccAWSCloudFormation_importBasic (85.19s)
=== RUN   TestAccAWSCloudFormation_basic
--- PASS: TestAccAWSCloudFormation_basic (83.02s)
=== RUN   TestAccAWSCloudFormation_yaml
--- PASS: TestAccAWSCloudFormation_yaml (83.35s)
=== RUN   TestAccAWSCloudFormation_defaultParams
--- PASS: TestAccAWSCloudFormation_defaultParams (83.42s)
=== RUN   TestAccAWSCloudFormation_allAttributes
--- PASS: TestAccAWSCloudFormation_allAttributes (125.01s)
=== RUN   TestAccAWSCloudFormation_withParams
--- PASS: TestAccAWSCloudFormation_withParams (153.93s)
=== RUN   TestAccAWSCloudFormation_withUrl_withParams
--- PASS: TestAccAWSCloudFormation_withUrl_withParams (239.40s)
=== RUN   TestAccAWSCloudFormation_withUrl_withParams_withYaml
--- PASS: TestAccAWSCloudFormation_withUrl_withParams_withYaml (129.49s)
=== RUN   TestAccAWSCloudFormation_withUrl_withParams_noUpdate
--- PASS: TestAccAWSCloudFormation_withUrl_withParams_noUpdate (180.77s)

=== RUN   TestAccAWSElasticSearchDomain_basic
--- PASS: TestAccAWSElasticSearchDomain_basic (935.70s)
=== RUN   TestAccAWSElasticSearchDomain_duplicate
--- PASS: TestAccAWSElasticSearchDomain_duplicate (590.45s)
=== RUN   TestAccAWSElasticSearchDomain_importBasic
--- PASS: TestAccAWSElasticSearchDomain_importBasic (1045.02s)
=== RUN   TestAccAWSElasticSearchDomain_v23
--- PASS: TestAccAWSElasticSearchDomain_v23 (979.27s)
=== RUN   TestAccAWSElasticSearchDomain_complex
--- PASS: TestAccAWSElasticSearchDomain_complex (1228.28s)

$ TF_ACC=1 go test ./aws -v -run=TestAccAWSGlacierVault_ -timeout 120m
=== RUN   TestAccAWSGlacierVault_importBasic
--- PASS: TestAccAWSGlacierVault_importBasic (17.45s)
=== RUN   TestAccAWSGlacierVault_basic
--- PASS: TestAccAWSGlacierVault_basic (13.90s)
=== RUN   TestAccAWSGlacierVault_full
--- PASS: TestAccAWSGlacierVault_full (13.89s)
=== RUN   TestAccAWSGlacierVault_RemoveNotifications
--- PASS: TestAccAWSGlacierVault_RemoveNotifications (25.02s)

$ TF_ACC=1 go test ./aws -v -run=TestAccAWSKmsKey_ -timeout 120m  
=== RUN   TestAccAWSKmsKey_importBasic
--- PASS: TestAccAWSKmsKey_importBasic (64.26s)
=== RUN   TestAccAWSKmsKey_basic
--- PASS: TestAccAWSKmsKey_basic (77.81s)
=== RUN   TestAccAWSKmsKey_disappears
--- PASS: TestAccAWSKmsKey_disappears (31.09s)
=== RUN   TestAccAWSKmsKey_policy
--- PASS: TestAccAWSKmsKey_policy (60.00s)
=== RUN   TestAccAWSKmsKey_isEnabled
--- PASS: TestAccAWSKmsKey_isEnabled (463.94s)
=== RUN   TestAccAWSKmsKey_tags
--- PASS: TestAccAWSKmsKey_tags (58.08s)
...
```

### CODE REVIEW

File: aws/structure_test.go [hunk 1]
```diff
@@ -1199,70 +1199,6 @@ func TestFlattenPolicyAttributes(t *testing.T) {
...
-	if err != nil {
-		t.Fatalf("Expected not to throw an error while parsing JSON, but got: %s", err)
-	}
-
-	if actual != expected {
-		t.Fatalf("Got:\n\n%s\n\nExpected:\n\n%s\n", actual, expected)
-	}
-
-	// Well formatted but not valid,
-	// missing closing squre bracket.
```
Comment (radeksimko) [inline aws/structure_test.go]: Do you mind just PRing this part of the test upstream? https://github.com/hashicorp/terraform/blob/master/helper/structure/normalize_json_test.go

File: aws/structure_test.go [hunk 2]
```diff
@@ -1199,70 +1199,6 @@ func TestFlattenPolicyAttributes(t *testing.T) {
...
-
-	// We expect the invalid JSON to be shown back to us again.
-	if actual != invalidJson {
-		t.Fatalf("Got:\n\n%s\n\nExpected:\n\n%s\n", actual, invalidJson)
-	}
-
-	// Verify that it leaves strings alone
-	testString := "2016-07-28t04:07:02z\nsomething else"
-	expected = "2016-07-28t04:07:02z\nsomething else"
-	actual, err = normalizeJsonString(testString)
```
Comment (radeksimko) [inline aws/structure_test.go]: Likewise - do you mind PRing this upstream? https://github.com/hashicorp/terraform/blob/master/helper/structure/normalize_json_test.go

File: aws/structure_test.go (hunk 1 above)
Comment (Ninir) [reply aws/structure_test.go]: Sure, here we go: https://github.com/hashicorp/terraform/pull/17122

File: aws/structure_test.go (hunk 2 above)
Comment (Ninir) [reply aws/structure_test.go]: Thanks for the catch, it has been accepted upstream. Merging!

**Pull request merged**